### DISCUSSION TITLE
Virtual Machine Support - KVM UI

### DISCUSSION THREAD

**Discussion by blakeseufert** [IceWhaleTech/CasaOS#1328] [Ideas]
It would be great to leverage the built in KVM funtions of linux virtualisation. This topic has come up with the most engagement in the last 2 CasaOS live streams (community events in Discord).

#### Describe the solution you'd like:
A way to create, update and destory virtual machines. A simple UI for setting up ram, cpu, disk images & connecting to noVNC. If possible in the future do some PCIE passthrough for extending VM's, adding gfx cards etc. This would enable greater flexibilty to run full servers, other OS's, Windows jump boxes, etc, etc.

#### Describe alternatives you've considered:
Unraid has a pretty decent solution for this but it's 2 biggest drawbacks are: 
1. Quite technical, need to know a lot about how a computer works to troubleshoot issues (OK for us, but CasaOS could be more mass market friendly, perhaps even download popular iso's as part of the setup to streamline it for tinkers.)
2. You need to create a parity drive when sometimes i don't need storage just some basic services running in a VM.

I know this is a hotly requested issues, yet have not been able to find mention of it.

Thanks for all your great works so far! 🙏

**Comment by TommySharpNZ** [comment]
Yeah making it easy to creat, run and manage virtual machines in CasaOS would be nice!

**Comment by AuthorShin** [comment]
I also asked for the same thing over here: https://github.com/IceWhaleTech/CasaOS/discussions/1288
Hope they'll agree on supporting it so we can start working on building it.

**Comment by nuelgoz** [comment]
Hello, it would be good to be able to implement Cloud Hypervisor, which is a virtual machine hypervisor written in Rust. It is compatible with Kata Containers and KVM. What I'm not sure about is whether it has a GUI for easy integration.

https://www.cloudhypervisor.org/

**Comment by eleazar7** [comment]
Yes!! please! I need this VM so I can run VM OS for content creation and gaming.  
i believe that the personal Server should replace everything in your home setup. everything is grouped into one machine!

**Comment by a7medo778** [comment]
yes that would be peeerfect,
the vm feature in synology is the only thingy keeping me latched to it

**Comment by paidlowered** [comment]
Running Ubuntu Server and I need a Debian VM for Home assistant supervised. Need mosquitto really badly for my Govee strips.

**Reply by mystikxkemix** [reply]
I'm using CasaOS for my HomeAssistant. HA is in a docker container, and Mosquitto is in a different container. That's working perfectly for some Tasmota devices 👍

**Comment by mickey3721** [comment]
Yes, some application need KVM to run.
I try to install KVM manager in base system, but conflict with CasaOS, virtual network not working well, ports always keep occupied by someone

**Reply by Gurulu-pc** [reply]
This! I'm not knowledgeable enough in networking to figure this out. I was trying to migrate my home assistant from my rpi3 to the more capable computer I use for casaOS essentially trying to self host all those more and more expansive cloud services. I disable my bridge config and moved everything back to the rpi because both host and guest just stopped communicating with the network randomly.
If you figure out how to properly configure network settings for KVM to work side by side with CasaOS, please share!
